### PR TITLE
DR 1108 soft deletes test

### PR DESCRIPTION
**Pull request by RoriCremer**

### CODE REVIEW

File: datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        repositoryApi.ingestDataset(datasetSummaryModel.getId(), ingestRequest);
+
+    ingestTabularDataJobResponse =
+        DataRepoUtils.waitForJobToFinish(repositoryApi, ingestTabularDataJobResponse);
+    IngestResponseModel ingestResponse =
+        DataRepoUtils.expectJobSuccess(
+            repositoryApi, ingestTabularDataJobResponse, IngestResponseModel.class);
+    logger.info("Successfully loaded data into dataset: {}", ingestResponse.getDataset());
+
+    String datasetId = datasetSummaryModel.getId(); // TODO should these be in utils?
```
Comment (snf2ye) [inline datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java]: If you wanted to track this in a util, you could switch your lines 50-75 to use the bulkLoadSetup() method from BulkLoadUtils.  That info would then be tracked in the bulkLoadUtils, so you could call bulkLoadUtils.getDatasetId(). 
But, tbh, I think that would require a good bit of re-org, so not sure if that needs to/should happen in this ticket.

File: datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        StreamSupport.stream(result.getValues().spliterator(), false)
+            .map(fieldValues -> fieldValues.get(0).getStringValue())
+            .collect(Collectors.toList());
+
+    logger.info("Successfully retrieved row ids for table: {}", tableName);
+
+    // write to GCS
+    String csvFileName = FileUtils.randomizeName("this-too-better-pass") + ".csv";
+    String csvFileRefName = dirInCloud + "/" + csvFileName;
+    String gcsPath = FileUtils.createGcsPath(rowIds, csvFileRefName, testConfigGetIngestbucket);
```
Comment (snf2ye) [inline datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java]: For my education: Why do you have to do this manual write to GCS? Per our conversation yesterday, wouldn't the bulk file load put the file in gcs?

**Review by snf2ye — APPROVED** [review]
A couple of questions, but looks good to me.

File: datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java (hunk 1 above)
Comment (marikomedlock) [reply datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java]: Actually, I've refactored the setup and cleanup methods out of BulkLoadUtils, so please don't do that. Depending on whether I merge my stuff first, or you merge first, we can change this to use the new shared code in one of our PRs or a separate one.

File: datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    DatasetModel datasetModel = repositoryApi.retrieveDataset(datasetId);
+    String dataProject = datasetModel.getDataProject();
+    String tableName = datasetModel.getSchema().getTables().get(0).getName();
+
+    // get row ids for table  TODO combine SQL query w queryForFileRefs in DRSLookup
+    String tableRef =
+        String.format(
+            "`%s.%s.%s`",
+            dataProject,
+            "datarepo_" + datasetModel.getName(), // TODO const ni BQ utils
```
Comment (marikomedlock) [inline datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java]: Agreed on pulling out anything shared between the DRSLookup and this test into BQ utils.

File: datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java (hunk 2 above)
Comment (marikomedlock) [reply datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java]: I think this is the file that contains the rows to soft delete, not file data. So we have to put that file in GCS first before pointing Data Repo at it.

File: datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            .collect(Collectors.toList());
+
+    logger.info("Successfully retrieved row ids for table: {}", tableName);
+
+    // write to GCS
+    String csvFileName = FileUtils.randomizeName("this-too-better-pass") + ".csv";
+    String csvFileRefName = dirInCloud + "/" + csvFileName;
+    String gcsPath = FileUtils.createGcsPath(rowIds, csvFileRefName, testConfigGetIngestbucket);
+
+    // build the deletion request with pointers to the file with row ids to soft delete
```
Comment (marikomedlock) [inline datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java:168]: I think this file with the row ids to soft delete contains all the row ids in the table -- right? If so, then that means we could only run the user journey below once? If I've got that right, then I think this is a good test for soft deletes, but perhaps a slightly restructured version could be better for testing scaling of soft deletes. To test scaling (i.e. how many soft delete requests can we handle at a time) perhaps we could break this up to one soft delete request per row id? Then each user journey thread would try and soft delete one row id.

File: datarepo-clienttests/src/main/java/utils/FileUtils.java [hunk 5]
```diff
@@ -128,7 +131,28 @@ public static String createGsPath(
 
     // save a reference to the JSON file so we can delete it in cleanup()
     createdScratchFiles.add(fileRefName);
-    return "gs://" + testConfigGetIngestbucket + "/" + fileRefName;
+    return String.format("gs://%s/%s", testConfigGetIngestbucket, fileRefName);
+  }
+
+  public static String createGcsPath( // TODO combine this and above method
```
Comment (marikomedlock) [inline datarepo-clienttests/src/main/java/utils/FileUtils.java]: Agreed that combining this with the above method would be good code reuse. Perhaps you could write to a ByteArrayOutputStream or use a StringBuilder and then get the bytes to pass to the above method from that.

**Review by marikomedlock — COMMENTED** [review]
This looks great, thanks for figuring out all the soft delete stuff! A few comments inline.

File: datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java (hunk 4 above)
Comment (RoriCremer) [reply datarepo-clienttests/src/main/java/testscripts/SoftDeleteDataset.java:168]: follow up pr with this additional test / work to come!

**Review by fboulnois — APPROVED** [review]
I agree with the earlier suggestions, otherwise lgtm.

**Pull request merged**